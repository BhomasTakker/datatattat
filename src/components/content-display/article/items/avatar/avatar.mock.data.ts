export const META_IMAGE =
	"https://th.bing.com/th?id=ORMS.0c195456aa1d54b17938c130df5cecc9&pid=Wdp&w=612&h=304&qlt=90&c=1&rs=1&dpr=1&p=0";

export const AVATAR_MOCK = {
	img: "https://e3.365dm.com/23/10/70x70/skynews-dwayne-johnson-waxwork_6333809.jpg?20231024132712",
	alt: "Dwayne Johnson 'belly laughs' after controversial waxwork unveiled",
	src: "https://news.sky.com/story/dwayne-johnson-belly-laughs-after-controversial-waxwork-unveiled-12991640",
};
