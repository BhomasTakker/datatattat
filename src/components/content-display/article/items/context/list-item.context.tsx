import { ReactNode, createContext, useCallback, useState } from "react";
import { useMeta } from "../hooks/useMeta";
import { CollectionItem } from "@/src/types/data-structures/collection/item/item";

type ListItemState = {
	item: CollectionItem;
};
// put me in types
interface Meta {
	image: string;
}

type ListItemInterface = {
	getMeta: () => void;
	meta: Meta | null;
};

const initialState: ListItemState & ListItemInterface = {
	item: {
		title: "",
		src: "",
		description: "",
		guid: "",
		variant: "",
	},
	getMeta: () => {},
	meta: {
		image: "",
	},
};

export const ListItemContextProvider = ({
	value,
	children,
}: {
	value: ListItemState;
	children: ReactNode;
}) => {
	const { item } = value;
	const { src } = item;
	const [needMeta, setNeedMeta] = useState(false);
	const { meta } = useMeta(src, needMeta);

	const getMeta = useCallback(() => {
		// console.log("FEATURE:0001", "GET:META", { src });
		setNeedMeta(true);
	}, []);
	return (
		// Would you always spread given value here?
		<ListItemContext.Provider value={{ ...value, getMeta, meta }}>
			{children}
		</ListItemContext.Provider>
	);
};

export const ListItemContext = createContext({ ...initialState });
