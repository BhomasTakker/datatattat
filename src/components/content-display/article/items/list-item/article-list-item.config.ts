// Don't like having getConfig here and continuiously being made

import { ListItemProps } from "@mui/material";
import { ViewportSize } from "../../card/article-card.wrapper";
import { ArticleListItemProps } from "./aticle-list-item";
type ConfigType = Omit<ArticleListItemProps, "item"> & ListItemProps;
export type ArticleListItemVariant = "compact" | "expanded";

// this likely needs a think
// Either props AND variant
// or an impossible number of variants
// Some of these props being defaults
// So compact - but maxLines, showAvatar, publisher, published, or author
const compact: ConfigType = {
	direction: "row",
	avatarSize: "sm",
	titleMaxLines: 1,
	useAvatar: true,
	showDescription: false,
	descriptionMaxLines: 1,
	showPublished: true,
	showAuthor: false,
	showPublisher: false,
	// divider: true,
	// disableGutters: true,
	// disablePadding: true,
	sx: { paddingY: 0 },
};

const expanded: ConfigType = {
	direction: "column",
	avatarSize: "md",
	titleMaxLines: 1,
	useAvatar: true,
	showDescription: true,
	descriptionMaxLines: 3,
	showPublished: true,
	showAuthor: false,
	showPublisher: true,
	// divider: true,
};

const getVariant = new Map<string, ConfigType>([
	["compact", compact],
	["expanded", expanded],
]);
// Have a set naming conventionand generic get
export const getArticleListItemConfig = (
	variant: ArticleListItemVariant,
	size?: ViewportSize
) => {
	return getVariant.get(variant) || expanded;
};
