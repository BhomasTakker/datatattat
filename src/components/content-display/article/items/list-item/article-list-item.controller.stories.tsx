// https://storybook.js.org/docs/react/writing-stories/introduction
import type { Meta, StoryObj } from "@storybook/react";
import { Article1 } from "../../mock/Articles.mock";
import { ArticleListItemController } from "./article-list-item.controller";

const meta: Meta<typeof ArticleListItemController> = {
	component: ArticleListItemController,
};
export default meta;
type Story = StoryObj<typeof ArticleListItemController>;

export const Controller: Story = {
	name: "Article List Item Controller",
	args: {
		item: Article1,
		variant: "expanded",
	},
};
