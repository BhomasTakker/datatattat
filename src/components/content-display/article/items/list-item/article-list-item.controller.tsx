import { CollectionItem } from "@/src/types/data-structures/collection/item/item";
import { ArticleListItem } from "./aticle-list-item";
import {
	ArticleListItemVariant,
	getArticleListItemConfig,
} from "./article-list-item.config";

export interface ArticleListItemControllerProps {
	item: CollectionItem;
	variant: ArticleListItemVariant;
}

// effectively pass in config list
export const ArticleListItemController = ({
	item,
	variant,
}: ArticleListItemControllerProps) => {
	const props = getArticleListItemConfig(variant);

	return <ArticleListItem {...props} item={item} />;
};
