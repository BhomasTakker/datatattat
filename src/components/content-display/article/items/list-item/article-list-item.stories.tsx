// https://storybook.js.org/docs/react/writing-stories/introduction
import type { Meta, StoryObj } from "@storybook/react";
import { ArticleListItem } from "./aticle-list-item";
import { Article1 } from "../../mock/Articles.mock";

const meta: Meta<typeof ArticleListItem> = {
	component: ArticleListItem,
};
export default meta;
type Story = StoryObj<typeof ArticleListItem>;

export const Primary: Story = {
	name: "Article List Item",
	args: {
		item: Article1,
		direction: "column",
		titleMaxLines: 1,
		useAvatar: true,
		showDescription: true,
		descriptionMaxLines: 3,
		showPublished: true,
		showAuthor: true,
		showPublisher: true,
	},
};
