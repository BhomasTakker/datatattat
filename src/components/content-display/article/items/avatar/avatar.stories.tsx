import type { Meta, StoryObj } from "@storybook/react";
import { ArticleAvatar } from "./avatar";

const meta: Meta<typeof ArticleAvatar> = {
	component: ArticleAvatar,
};
export default meta;
type Story = StoryObj<typeof ArticleAvatar>;

export const Standard: Story = {
	args: {
		alt: "My Alt Text",
		img: "https://e3.365dm.com/23/10/70x70/skynews-dwayne-johnson-waxwork_6333809.jpg?20231024132712",
		src: "https://news.sky.com/story/dwayne-johnson-belly-laughs-after-controversial-waxwork-unveiled-12991640",
	},
};
