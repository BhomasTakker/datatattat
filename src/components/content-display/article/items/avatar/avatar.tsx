import { Avatar } from "@mui/material";
import { useContext } from "react";
import { ListItemContext } from "../context/list-item.context";

interface ArticleAvatar {
	alt: string;
	img: string;
	src: string;
}

// Maybe neaten this up
// get data from context
export const ArticleAvatar = ({ alt, img, src }: ArticleAvatar) => {
	const { getMeta, meta } = useContext(ListItemContext);
	const { image } = meta || {};

	if (!img && !image) {
		getMeta();
		return <></>;
	}

	return <Avatar alt={alt} src={img || image} />;
};
