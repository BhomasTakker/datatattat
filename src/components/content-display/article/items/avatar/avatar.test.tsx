import { render, screen } from "@testing-library/react";
import { ArticleAvatar } from "./avatar";
import { AVATAR_MOCK } from "./avatar.mock.data";
import { ListItemContextProvider } from "../context/list-item.context";
import { Article1 } from "../../mock/Articles.mock";

const mockFunc = jest.fn();

jest.mock("../hooks/useMeta", () => ({
	useMeta: () => mockFunc,
}));

const renderWithContext = (props: {
	alt: string;
	img: string;
	src: string;
}) => {
	render(
		<ListItemContextProvider value={{ item: Article1 }}>
			<ArticleAvatar {...props} />
		</ListItemContextProvider>
	);
};

describe("Article Avatar", () => {
	it("should render an ArticleAvatar ", () => {
		renderWithContext(AVATAR_MOCK);
		const avatarElement = screen.getByTestId("avatar");
		expect(avatarElement).toBeInTheDocument();
	});

	it("should render an img ", () => {
		renderWithContext(AVATAR_MOCK);
		const avatarElement = screen.getByRole("img");
		expect(avatarElement).toBeInTheDocument();
	});

	it("should render expected alt ", () => {
		renderWithContext(AVATAR_MOCK);
		const avatarElement = screen.getByAltText(AVATAR_MOCK.alt);
		expect(avatarElement).toBeInTheDocument();
	});

	it("should render image with expected src", () => {
		renderWithContext(AVATAR_MOCK);
		const avatarElement = screen.getByRole("img");
		expect(avatarElement).toHaveAttribute("src", AVATAR_MOCK.img);
	});

	it("should render empty fragment if no image", async () => {
		renderWithContext({ ...AVATAR_MOCK, img: "" });
		const avatarElement = screen.queryByTestId("avatar");
		expect(avatarElement).not.toBeInTheDocument();
	});

	describe("snaps", () => {
		it("should match snapshot", () => {
			renderWithContext(AVATAR_MOCK);
			const avatarElement = screen.getByTestId("avatar");
			expect(avatarElement).toMatchSnapshot();
		});
	});

	// describe("useMeta", () => {
	// we are rendering with context but no testing
	// or mocking of that context etc
	// so what is it doing?
	// we can render without context
	// but then we won't be testing all possibilities
	// });
});
