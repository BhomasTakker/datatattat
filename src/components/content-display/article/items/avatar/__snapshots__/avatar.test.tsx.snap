// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Avatar snaps should match snapshot 1`] = `
<div
  class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
  data-testid="avatar"
>
  <img
    alt="Dwayne Johnson 'belly laughs' after controversial waxwork unveiled"
    class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
    src="https://e3.365dm.com/23/10/70x70/skynews-dwayne-johnson-waxwork_6333809.jpg?20231024132712"
  />
</div>
`;
